smi/adaptor/controller.go

```diff
@@ -0,0 +1,340 @@
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		log.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the Ts resource with this namespace/name
+	ts, err := c.tsLister.TrafficSplits(namespace).Get(name)
+	if err != nil {
+		// Check if TS does not exit anymore
+		if errors.IsNotFound(err) {
+			// Check if there is a relevant SP that was created or updated by SMI Controller
+			selector := labels.NewSelector()
+			r, err := labels.NewRequirement(createdOrUpdatedBySMIAdaptorFor, selection.Equals, []string{name})
+			if err != nil {
+				return err
+			}
+			selector.Add(*r)
+			sps, err := c.spLister.ServiceProfiles(namespace).List(selector)
+			if err != nil {
+				return err
+			}
+
+			if len(sps) != 0 {
+				// Empty dstOverrides in the SP
+				sp := sps[0]
+				if ignoreAnnotationPresent(sp) {
+					log.Infof("skipping %s sp as ignore annotation is present", sp.Name)
+					return nil
+				}
+
+				sp.Spec.DstOverrides = nil
+				_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(namespace).Update(context.Background(), sp, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		return err
+	}
+
+	// Check if the Service Profile is already Present
+	sp, err := c.spLister.ServiceProfiles(ts.Namespace).Get(fmt.Sprintf("%s.%s.svc.%s", ts.Spec.Service, ts.Namespace, c.clusterDomain))
+	if err != nil {
+		// Create a Service Profile resource as it does not exist
+		_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Create(context.Background(), c.toServiceProfile(ts), metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Infof("Service Profile already present")
+		if ignoreAnnotationPresent(sp) {
+			log.Infof("skipping %s sp as ignore annotation is present", sp.Name)
+			return nil
+		}
+
+		// Check if SP Matches the TS, and update if it not
+		spFromTs := c.toServiceProfile(ts)
+		if !equal(spFromTs, sp) {
+			log.Infof("serviceprofile/%s does not match trafficscplit/%s", sp.Name, ts.Name)
+			updateDstOverrides(sp, ts, c.clusterDomain)
+			log.Infof("Updated Service Profile as it's not equivalent to the relevant TS")
+			_, err = c.spclientset.LinkerdV1alpha2().ServiceProfiles(ts.Namespace).Update(context.Background(), sp, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func ignoreAnnotationPresent(sp *serviceprofile.ServiceProfile) bool {
+	_, ok := sp.Annotations[ignoreServiceProfileAnnotation]
+	return ok
+}
+
+func equal(spA *serviceprofile.ServiceProfile, spB *serviceprofile.ServiceProfile) bool {
+	if spA.Name != spB.Name {
+		return false
+	}
+
+	if spA.Namespace != spB.Namespace {
+		return false
+	}
+
+	if len(spA.Spec.DstOverrides) != len(spB.Spec.DstOverrides) {
+		return false
+	}
+
+	for index, dstA := range spA.Spec.DstOverrides {
```

Comment:
Out of curiosity, is there a guarantee that `DstOverrides` between `spA` and `spB` will have the same ordering if both have the same `len(DstOverrides)`?

Reply:
Totally Valid concern. Updated the PR to consider ordering here!